fdb-extensions/src/main/java/com/apple/foundationdb/util/LoggableKeysAndValues.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * LoggableKeysAndValues.java
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2015-2019 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.util;
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+
+/**
+ * Associates loggable information with an object as a map.
+ * Record layer log messages are (relatively) well formed. They are
+ * comprised of a static "title" of the message along with a set of
+ * keys and values that provide context about the "title" of the message.
+ * For example a "File not found" log entry may include keys and values
+ * like filename="foo" and directory="/bar". This makes the logs easy
+ * to generally search for all such errors, later extracting the set of
+ * files and directories that couldn't be found. This interface defines
+ * the methods that all objects that wish to provide such logging details
+ * must implement.
+ *
+ * @typeparam T type of object to associate loggable information with
+ */
+interface LoggableKeysAndValues<T> {
```

Comment:
Should this be `T extends LoggableKeysAndValues` to underscore that the implementors should probably supply their own type here? It probably doesn’t matter for, like, correctness, but it might be nice just for future implementors (and technically, adding it later would be a breaking change).

Reply:
LoggableKeysAndValues is the interface being defined here. There is no restriction on the type of class T that can implement the interface. Anyway, adding T extends LoggableKeysAndValues will cause build errors.

Reply:
Oh, I see. I think it's actually somewhat subtle, but is the problem that `LoggableKeysAndValues<T extends LoggableKeysAndValues>` complains about raw types?

Reply:
Yeah.